sky/check.py

```diff
@@ -20,44 +21,99 @@
-                if activated_account is not None:
-                    echo(f'    Activated account: {activated_account}')
-            if reason is not None:
-                echo(f'    Hint: {reason}')
+    def check_credentials(
+            cloud: Union[sky_clouds.Cloud, ModuleType],
+            capability: CloudCapability) -> Tuple[bool, Optional[str]]:
+        if capability == CloudCapability.COMPUTE:
+            return cloud.check_credentials()
+        elif capability == CloudCapability.STORAGE:
+            return cloud.check_storage_credentials()
+        else:
+            raise ValueError(f'Invalid capability: {capability}')
+
+    def get_cached_state(capability: CloudCapability) -> List[sky_clouds.Cloud]:
+        if capability == CloudCapability.COMPUTE:
+            return global_user_state.get_cached_enabled_clouds()
+        elif capability == CloudCapability.STORAGE:
+            return global_user_state.get_cached_enabled_storage_clouds()
         else:
-            disabled_clouds.append(cloud_repr)
-            echo(f'    Reason: {reason}')
+            raise ValueError(f'Invalid capability: {capability}')
+
+    def set_cached_state(clouds: List[str],
+                         capability: CloudCapability) -> None:
+        if capability == CloudCapability.COMPUTE:
+            global_user_state.set_enabled_clouds(clouds)
+        elif capability == CloudCapability.STORAGE:
+            global_user_state.set_enabled_storage_clouds(clouds)
+        else:
+            raise ValueError(f'Invalid capability: {capability}')
+
+    def check_one_cloud(cloud_tuple: Tuple[str, Union[sky_clouds.Cloud,
+                                                      ModuleType]],
+                        capabilities: List[CloudCapability]) -> None:
+        cloud_repr, cloud = cloud_tuple
+        for capability in capabilities:
+            not_supported = False
+            with rich_utils.safe_status(f'Checking {cloud_repr}...'):
+                try:
+                    ok, reason = check_credentials(cloud, capability)
+                except exceptions.NotSupportedError:
+                    not_supported = True
+                except Exception:  # pylint: disable=broad-except
+                    # Catch all exceptions to prevent a single cloud
+                    # from blocking the check for other clouds.
+                    ok, reason = False, traceback.format_exc()
+            if not_supported:
+                continue
+            status_msg = ('enabled' if ok else 'disabled')
+            styles = {'fg': 'green', 'bold': False} if ok else {'dim': True}
+            echo('  ' + click.style(f'{cloud_repr}: {status_msg}', **styles) +
+                 ' ' * 30)
+            if ok:
+                if cloud_repr not in enabled_clouds:
+                    enabled_clouds[cloud_repr] = [capability]
+                else:
+                    enabled_clouds[cloud_repr].append(capability)
```

Comment:
```suggestion
                enabled_clouds.setdefault(cloud_repr, []).append(capability)
```
not sure if this is more readable or more confusing, will leave it up to you

Reply:
TIL!